pkg/operator/controller/status.go

```diff
@@ -40,71 +46,153 @@ func (r *reconciler) syncOperatorStatus() error {
+	if err := r.client.List(context.TODO(), ingressList, client.InNamespace(r.Namespace)); err != nil {
+		return nil, fmt.Errorf("failed to list IngressControllers: %v", err)
 	}
 
-	return nil
+	return ingressList.Items, nil
 }
 
-// getOperatorState gets and returns the resources necessary to compute the
-// operator's current state.
-func (r *reconciler) getOperatorState() (*corev1.Namespace, []operatorv1.IngressController, error) {
-	ns := manifests.RouterNamespace()
+// computeOperatorStatusVersions computes the operator's current versions and also updates progressing condition if needed.
+func (r *reconciler) computeOperatorStatusVersions(oldVersions []configv1.OperandVersion, conditions []configv1.ClusterOperatorStatusCondition) ([]configv1.OperandVersion, []configv1.ClusterOperatorStatusCondition) {
+	oldReleaseVersion := UnknownVersionValue
+	oldIngressControllerImage := UnknownVersionValue
+	for _, opv := range oldVersions {
+		switch opv.Name {
+		case OperatorVersionName:
+			oldReleaseVersion = opv.Version
+		case IngressControllerVersionName:
+			oldIngressControllerImage = opv.Version
+		}
+	}
 
-	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
-		if errors.IsNotFound(err) {
-			return nil, nil, nil
+	operatorAvailable := false
+	operatorProgressing := false
+	var progressingCondition *configv1.ClusterOperatorStatusCondition
+	for i, cond := range conditions {
+		switch cond.Type {
+		case configv1.OperatorAvailable:
+			if cond.Status == configv1.ConditionTrue {
+				operatorAvailable = true
+			}
+		case configv1.OperatorProgressing:
+			progressingCondition = &conditions[i]
+			if cond.Status == configv1.ConditionTrue {
+				operatorProgressing = true
+			}
 		}
+	}
 
-		return nil, nil, fmt.Errorf(
-			"error getting Namespace %s: %v", ns.Name, err)
+	releaseVersion := r.OperatorReleaseVersion
+	ingressControllerImage := r.IngressControllerImage
+	// We need to report old version until the operator fully transitions to the new version.
+	// https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#version-reporting-during-an-upgrade
+	if !operatorAvailable || operatorProgressing {
+		releaseVersion = oldReleaseVersion
+		ingressControllerImage = oldIngressControllerImage
 	}
 
-	ingressList := &operatorv1.IngressControllerList{}
-	if err := r.client.List(context.TODO(), ingressList, client.InNamespace(r.Namespace)); err != nil {
-		return nil, nil, fmt.Errorf("failed to list IngressControllers: %v", err)
+	// Update progressing condition if needed
+	if (releaseVersion != r.OperatorReleaseVersion) || (ingressControllerImage != r.IngressControllerImage) {
+		if !operatorProgressing {
+			progressingCondition.Status = configv1.ConditionTrue
+			progressingCondition.Reason = "Reconciling"
+		}
+		if releaseVersion != r.OperatorReleaseVersion {
+			if len(progressingCondition.Message) != 0 {
+				progressingCondition.Message += " "
```

Comment:
IMO the progressing message can simply be updated instead of appending the new message to the old message. Thoughts?

Reply:
I thought by appending the message, we are providing accurate state noticed by the controller.
Depending on the circumstances, message may have one ore more of these strings:
- X ingress controllers available, want Y
- Moving to release version X
- Moving to ingress-controller image version Y

Reply:
What keeps the operator from repeatedly appending the same "Moving to [...]" strings to the condition?

Reply:
We create progressing condition from scratch in every status iteration and this eliminates appending of same "Moving to" messages (https://github.com/openshift/cluster-ingress-operator/blob/f09e8ee06d8f9bebb0e5aed94cbbe4a41e3eee38/pkg/operator/controller/status.go#L231).

Reply:
Ah, because `computeOperatorProgressingCondition` resets the message.  Got it.

Reply:
For the string building, you could start with a `messages []string`, append to it as you go, and at the end `condition.Messages = strings.Join(messages, "\n")`

Reply:
done